### ISSUE TITLE
Engine start writes user config

### ISSUE THREAD

**Issue by morganwillcock** [adventuregamestudio/ags#1156]
Just running a game writes a user configuration file, which then overrides any change made to the game's configuration file.

This is extremely confusing since:
* editing the game configuration (e.g. windowed mode) would work before starting the game
* editing the game configuration (e.g. windowed mode) would not work after the game was started once

```
$ cat /home/mwillcock/.local/share/ags/DaggerHell/acsetup.cfg
cat: /home/mwillcock/.local/share/ags/DaggerHell/acsetup.cfg: No such file or directory

$ ~/Downloads/DAGGERHELL/DAGGERHELL
Adventure Game Studio v3.5 Interpreter
Copyright (c) 1999-2011 Chris Jones and 2011-2020 others
ACI version 3.5.0.27

Initializing allegro
Initializing game data
Located game data file: /home/mwillcock/Downloads/DAGGERHELL/data/DaggerHell.ags
Opened game data file: game28.dta
Game data version: 49
Compiled with: 3.4.3.1
Setting up game configuration
Data directory: /home/mwillcock/Downloads/DAGGERHELL/data/
Setting up window
Initializing TTF renderer
Initializing keyboard
Initializing mouse: number of buttons reported is 3
Install timer
Sound settings: digital driver ID: 'AUTO' (0xffffffff), MIDI driver ID: 'AUTO' (0xffffffff)
Trying to init: digital driver ID: 'SDL2' (0x53444c32), MIDI driver ID: 'DIGI' (0x44494749)
Installed digital driver ID: 'SDL2' (0x53444c32), MIDI driver ID: 'DIGI' (0x44494749)
Install exit handler
Initialize legacy path finder library
Game GUI version: 118
Requested script API: v3.4.1 (5), compat level: v3.4.1 (5)
Game title: 'DaggerHell'
Checking for disk space
Initializing MOD/XM player
Game native resolution: 320 x 200 (32 bit)
Graphic settings: driver: D3D9, windowed: yes, screen def: max, screen size: 0 x 0, match device ratio: ignore, game scale: max_round
Requested graphics driver 'D3D9' not found, will try existing drivers instead
Running OpenGL: 3.0
Mouse control: off, base: 1.000000, speed: 1.000000
Initialize sprites
Audio is processed on the main thread
Engine initialization complete
Starting game
***** ENGINE HAS SHUTDOWN

$ cat /home/mwillcock/.local/share/ags/DaggerHell/acsetup.cfg
[graphics]
render_at_screenres=0
windowed=1
[language]
translation=
[mouse]
control_enabled=1
speed=1.000000
```

**Comment by ivan-mogilko** [comment]
> * editing the game configuration (e.g. windowed mode) would not work after the game was started once

I take it you were editing config file in the game directory?

This is the kind of question that was raised several times before, both in the trackers and on forums.
These are current related tickets: #975, #961.
This is a topic I opened on forums to ask opinions on the subject: https://www.adventuregamestudio.co.uk/forums/index.php?topic=58644.0

There are two things here:
* engine writing last configuration that may be changed at runtime. Is it beneficial or not, should it not allowed to be doing this?
* user config not being in the same directory, which may result in wrong assumptions on editing default one.

**Comment by morganwillcock** [comment]
> I take it you were editing config file in the game directory?

I am trying to write a setup program using the output of `--tell`.

> This is a topic I opened on forums to ask opinions on the subject: https://www.adventuregamestudio.co.uk/forums/index.php?topic=58644.0

I did see this but was a little worried that the feedback you will get there be too specific to certain use cases, because it is a complicated topic. Plus there seems to be fundamental problems with the actual settings themselves because there is no mechanism to deal with older settings being phased out, merged, or upgraded.

Fundamentally I think it should be managed in the same way as save data, so the engine should start with the safest defaults and then try to use/upgrade any existing settings. Settings downgrade is the same problem as save game downgrade.

If we are just talking about locations on disk to use, I would think It is probably beneficial to try and gather a list of the previous issues which have caused people problems, and assess whether these were actual problems that a game engine needs to have solutions for, or whether it was a workaround for some self-inflicted AGS-ism like `RunAGSGame`.

> engine writing last configuration that may be changed at runtime. Is it beneficial or not, should it not be allowed to be doing this? How else to sync e.g. setup outside and setup done inside the game?

It probably is beneficial, but I didn't change any configuration whilst the game was running, so I wouldn't have expected anything to be written back.

> user config not being in the same directory, which may result in wrong assumptions on editing default one.

In this instance I was aware of where the user config would be.

**Comment by ivan-mogilko** [comment]
> I am trying to write a setup program using the output of `--tell`.

Guess I don't understand full situation then. You said that "editing the game configuration would not work after the game was started once", which file does your setup program edit? Or does "--tell" command not give correct instructions of where user config is supposed to be located?

> there is no mechanism to deal with older settings being phased out, merged, or upgraded.

Well, there is deprecated setting processing in the engine, because someone asked me to do this, but I could skip something too.
https://github.com/adventuregamestudio/ags/blob/master/Engine/main/config.cpp#L383

> It probably is beneficial, but I didn't change any configuration whilst the game was running, so I wouldn't have expected anything to be written back.

I could change it to compare settings with defaults and not write user config, (it would have to both test whether user config exists and if latest settings are different), I just don't know if it's worth it.


> If we are just talking about locations on disk to use, I would think It is probably beneficial to try and gather a list of the previous issues which have caused people problems

Maybe I forgot to cover this in detail in that particular post, but notable problems are:
* not all game locations are writeable, this is why "user config" was split out and saved elsewhere.
* this makes it difficult to realize how to edit config by hand, because editing file in the game dir does not always lead to assumed results.
* some people simply don't like it that the games write files elsewhere and pollute obscure locations on their disk (that's to paraphrase what users told me).

RunAGSGame is a special situation, as each game in theory has its own config. So long as you start same program first all collection will use same config (because it's not reloaded when another game is loaded), but that's more of a side effect of how engine works.

**Comment by morganwillcock** [comment]
> Guess I don't understand full situation then. You said that "editing the game configuration would not work after the game was started once", which file does your setup program edit?

It is selectable which file will be edited, defaulting to the user config file.

> Or does "--tell" command not give correct information?

So far the `--tell` information has always been correct.

> > there is no mechanism to deal with older settings being phased out, merged, or upgraded.
> 
> Well, there is deprecated setting processing in the engine, because someone asked me to do this, but I could miss some settings too.
> https://github.com/adventuregamestudio/ags/blob/master/Engine/main/config.cpp#L383

As far as I can tell `control_enabled` for `[mouse]` is effectively replaced by `control_when`.

> Maybe I forgot to cover this in detail in that particular post, but notable problems are:
> 
>     not all game locations are writeable, this is why "user config" was split out and saved elsewhere.

I don't think anyone can/would argue against the reasoning, because game installation is outside the scope of the engine. The current per-user location on Windows is not the 'correct' location, but the concept of storing the data per user and outside of the game folder isn't really debatable, that is how programs on Windows are meant to work. Really the discussion is how far to go in accommodating requirements that fall outside of this, and that depends on what the expectations of the engine itself is and where it is expected to run.

Looking at an engine with similar scope (https://www.renpy.org/doc/html/config.html#var-config.save_directory) there are three options:

* Default to a sensible per user location
* Optionally use a custom path
* Optionally use the game directory (with a warning not to do it)

>     this makes it difficult to realize how to edit config by hand, because editing file in the game dir does not always lead to assumed results.

To ask the reverse question, if a game creator supplies default settings and then any user specific changes go to another location (say one of three from the renpy list, defined by the game) can anyone list some reasons why they would need to edit the config file that comes with the game? As it stands, winsetup follows these principles. Maybe for fiddling with old games running on newer engine versions, but I think that is outside of what the engine is primarily meant to be for.

>     some people simply don't like it that the games write files elsewhere and pollute obscure locations on their disk (that's to paraphrase what users told me). There is also no automated cleanup as most "homemade" games don't have proper installers, so these files may stack up there for years (same goes to saves tbh).

I can understand that, but there are standard locations which are meant to be used for such data, and this problem isn't unique to AGS games.

**Comment by rofl0r** [comment]
>    Default to a sensible per user location
>    Optionally use a custom path
>    Optionally use the game directory (with a warning not to do it)

my preference would be to always use game directory if possible (means there is write access), otherwise the .local/share/whatever path. optionally there could be a file (on posix-compatibles) /etc/ags.path which contains the user's preference, so one doesn't have to add a --command-line-flag whenever ags engine is started (although theoretically one could just write a wrapper script that does

```
#!/bin/sh
ags --save-to-game-directory "$@"
```
)

**Comment by morganwillcock** [comment]
> my preference would be to always use game directory if possible (means there is write access)

You cannot test against failure for this on Windows because writes may or may not be redirected back into the user profile, depending on where the non-writeable game directory is.

**Comment by ericoporto** [comment]
As far as I understood this ticket is about why  running the engine writes a config when the game runs and not changing directories right? And the suggestion is don't write config files - unless specifically told so maybe, I think windowed to fullscreen change is written on the config?

**Comment by morganwillcock** [comment]
> As far as I understood this ticket is about why running the engine writes a config when the game runs and not changing directories right? And the suggestion is don't write config files - unless specifically told so maybe, I think windowed to fullscreen change is written on the config?

Yes. Unless the settings are actually changed whilst the game is running it creates additional confusion to just to write them back to the user configuration.

**Comment by ivan-mogilko** [comment]
What classifies as the settings change inside the engine? For instance, player may press alt+enter to switch from fullscreen to windowed, will that classify as an option change, or it has to be something more explicit?

**Comment by ivan-mogilko** [comment]
I also have concern of other kind of situation. Imagine player runs the game without changing anything and that suits the needs. Then player installs game update (or there's an autoupdate if it's a Steam game for instance), which includes default config file, and suddenly some settings are changed, and game runs differently than prior. Would not that also count as unexpected behavior? I realize this is a rare occurence, but theoretically possible. This is also one of the reasons I was considering having 2 separate filenames for default and user config (literally default.cfg and user.cfg), which could solve this issue if user config is stored in same directory.

**Comment by ericoporto** [comment]
> Also, frankly, I still don't understand general idea. Why is it confusing that engine creates user config on the first run?

I don't think it's confusing too and it's completely acceptable behavior, this could be simply tagged won't fix and closed. I imagine a lot of games and software work exactly like this. Alternatively, just documenting the behavior.

---

I will try to follow the issue at hand to see if I understand what is written here, looking at the specific config files generated:
```
[graphics]
render_at_screenres=0
windowed=1
[language]
translation=
[mouse]
control_enabled=1
speed=1.000000
```

The changes that are written are changes that are accessible by the Script API. 

Should this be written in the file if the specific setting is changed at runtime? I think it makes sense that last behavior is the default behavior in the next run. 

So maybe the issue is we only want last "user selected" behavior? Does DaggerHell changes any of these configs in game, at runtime, that these specific settings should be written? I mean, it's a short game with no translation so it probably would not have any translation selection code. So these are probably written once the game exits whether player changed these in game or not (do I want to write to disk everytime I press Alt+Enter ? Probably not). 

So a behavior in my head for what I think this issue is about, would be, at proper exit (the game did not crash and user exited the game), the settings that were changed by the script API (marked with a flag somewhere), get written to the user config. This appears nice to have but not a priority.

Note that the changes in user config would not be overwritten by Steam because they are in other directory, so they are not affected by the game update. In the Steam case, user config should be never on the same directory as game config, since the game directory is not a place where anyone but steam should write files, otherwise configs are lost when the game updates. It's better to treat game directory as if it was a read only location, like in the CD days, since this will happen in case of games delivered by any commercial store, and some do use read only locations. The way AGS works is the same as all other games I have.

---

> I start to get a feeling that this confusion occurs not because engine does something, but because this behavior is not well documented, plus the setup tool or its user make wrong assumptions.

In my opinion, all of these other issues mentioned about config files are only a consequence of:
- There is no setup GUI in other OSes.
- The setup GUI in Windows is considered confusing.
- Some people use AGS Engine as if it was ScummVm, or a game player, which would need a setup GUI with all games in mind.

**Comment by ivan-mogilko** [comment]
> So a behavior in my head for what I think this issue is about, would be, at proper exit (the game did not crash and user exited the game), the settings that were changed by the script API (marked with a flag somewhere), get written to the user config. This appears nice to have but not a priority.

There's no need to flag every option that may be changed by API, that would create extra burden. The much simplier way is to create config tree in memory and then compare to the one on disk. This is practically what engine is doing, except it does not compare anything and just writes regardless.

**Comment by necros2k7** [comment]
Yes it really irritates that savegames and config files are saved at win\saved games\ path, I think optimal would be in game directory. It makes indie games portable, no pain in backing up savegames, makes drive more clean, winsetup.exe is always there if one likes to change settings.

**Comment by necros2k7** [comment]
So what`s your argument Erico?

**Comment by ericoporto** [comment]
This issue is NOT ABOUT storage location.

**Comment by necros2k7** [comment]
Ok, I got it, so anyway no point in creating file IMHO if nothing changed yet

**Comment by ivan-mogilko** [comment]
I decided to disable config saving upon game exit, at least temporarily, starting with the 3.5.0 patch.
The main reason is that it causes additional confusion for linux users, as they dont have a standard setup utility, so have to edit config by hand.

Currently engine is saving:
* windowed mode
* render sprites at screen res
* mouse control_enabled / speed
* translation

all of those may be edited by script, in which case game authors normally make custom config files anyway.

**Comment by ivan-mogilko** [comment]
@morganwillcock I disabled writing config in the engine for now (2c928a000418ac8de87a57155f36b4c39c3eca8b).
The main reason is that without proper solution for #975 and similar this lead to generally confusing behavior for users who prefered to edit game dir config by hand.

This behavior may still be implemented by game authors using custom files (as they were doing anyway, probably).

**Comment by morganwillcock** [comment]
OK, thanks!
